en/lessons/basics/collections.md

```diff
@@ -3,209 +3,774 @@ version: 1.3.1
+
+In that case our original example breaks down like so:
+
+```haskell
+ghci> [1,2,3,4]
+[1,2,3,4]
+ghci> 1 : 2 : 3 : 4 : []
+[1,2,3,4]
+ghci> Cons 1 (Cons 2 (Cons 3 (Cons 4 Nil)))
+Cons 1 (Cons 2 (Cons 3 (Cons 4 Nil)))
+```
+
+### List Performance
+
+Haskell implements lists as linked lists. Where the cons cells (this operator is
+called a cons `:`) act as the links.  This means that accessing the list length
+is an operation that will run in linear time (`O(n)`).  For this reason, it is
+typically faster to prepend than to append:
+
+```Haskell
+ghci> list = ["Orange", "Banana", "Apple"]
+["Orange", "Banana", "Apple"]
+-- Prepending (fast) (O(1))
+ghci> "Grape" : list
+["Grape", "Orange", "Banana", "Apple"]
+-- Appending (slow) (O(n))
+ghci> list ++ ["Cherry"]
+["Orange", "Banana", "Apple", "Cherry"]
+```
+
+### When to Use
+
+Linked lists are an extremely common data structure in functional programming,
+this means you are going to see them everywhere in Haskell. They are generally
+the first container a Haskell programmer reaches for.  Because of slow append
+and relatively slow indexed access (O(n) where n is the index) they are
+generally used in situations where you know you are going to have to iterate
+over the entire data set, or you want to preserve the order of elements.
+
+A great example of an abstract data structure for which a linked list is a great
+concrete implementation are stacks. This is because pushing and popping are
+O(1).
+
+A bad use case would be for a queue, where either enqueue or dequeue would be
+O(n) depending on which side of the linked list you decide to insert into.
+
+A common real world example of where lists are used is database queries. A
+database query could return no results `[]`, or some `[entity..]`, and there is
+potentially an ordering to these results. The database library doesn't really
+care about indexed access, and so it leaves that consideration to the caller of
+the function.
+
 ### List Concatenation
 
-List concatenation uses the `++/2` operator:
+List concatenation uses the `++` operator:
 
-```elixir
-iex> [1, 2] ++ [3, 4, 1]
+```haskell
+ghci> [1, 2] ++ [3, 4, 1]
 [1, 2, 3, 4, 1]
 ```
 
-A side note about the name (`++/2`) format used above:
-In Elixir (and Erlang, upon which Elixir is built), a function or operator name has two components: the name you give it (here `++`) and its _arity_.
-Arity is a core part of speaking about Elixir (and Erlang) code.
-It is the number of arguments a given function takes (two, in this case).
-Arity and the given name are combined with a slash. We'll talk more about this later; this knowledge will help you understand the notation for now.
+A side note about another operator that you might come across `<>`. This
+operator is sometimes referred to as `mappend`, it represents generic
+concatenation. In other words any type that supports `<>` can be "appended". In
+the case of lists `<>` is exactly `++` under the hood.
```

Comment:
Like at the beginning of the list lesson, or at the beginning of the containers section? I already received feedback to make this explanation more simple, and I think having it located next to list concatenation gives important context. But I could be convinced otherwise!